docs/proposals/thin-volumes.md

```diff
@@ -0,0 +1,175 @@
+
+## Proposal
+- Topolvm is **not** responsible for creating the thin-pools. Storage admin/operator has to create the thin-pools in advance while creating the volume groups.
+- Multiple device classes can be mapped to a single volume group having multiple thin pools. But each device class should map to a single thin pool within that volume group.
+- `lvmd.yaml` file should be updated to hold the mapping between device classes, volume group and thin pools.
+- For device classes mapped to thin pools, the node annotation `capacity.topolvm.cybozu.com/<device-class>` should be updated with respect to the thin pool size and overprovision ratio.
+- Metrics related to thin volume `data` and `metadata` parameters should be exported.
+
+
+
+## Design Details
+- A storage admin has to create the volume groups and the thin-pools on this volume group.
+- The storage admin should update the lvmd.yaml config file on the nodes to map device classes, volume groups and the corresponding thin-pools.
+- Sample `lvmd.yaml` config file:
+``` yaml=
+device-classes:
+  - name: ssd-thin
+    volume-group: myvg1
+    default: true
+    spare-gb: 10
+    type: thin
+    thin-pool:
+        name: pool0
+        overprovision-ratio: 50.0
+  - name: ssd
+    volume-group: myvg1
+    default: true
+    spare-gb: 10
+```
+- Above example shows two device classes; `ssd-thin` and `ssd`. Both use `myvg1` volume group.
+- `ssd-thin` has thin pool parameters.
+- Only thin volumes will be created when `ssd-thin` device class is selected by the user in the storageClass parameter `capacity.topolvm.cybozu.com/device-class`
+- Only thick volumes will be created when user has specified `ssd` in the storageClass.
+
+#### API Changes
+
+- The `DeviceClass` struct of the device-class-manager will hold the information about the thin-pool associated with the Volume Group.
+
+``` go
+
++ type ThinPoolConfig struct {
++   // Name of thin pool
++   Name string`json:"name"`
++   // OverProvisionRatio represents the ratio of overprovision that can be + allowed on thin pools
++   OverProvisionRatio float64 `json:"overprovision-ratio"`
++}
+
+type DeviceClass struct {
+    // Name for the device-class name
+    Name string `json:"name"`
+    // Volume group name for the device-class
+    VolumeGroup string `json:"volume-group"`
+    // Default is a flag to indicate whether the device-class is the default
+    Default bool `json:"default"`
+    // SpareGB is storage capacity in GiB to be spared
+    SpareGB *uint64 `json:"spare-gb"`
+    // Stripe is the number of stripes in the logical volume
+    Stripe *uint `json:"stripe"`
+    // StripeSize is the amount of data that is written to one device before moving to the next device
+    StripeSize string `json:"stripe-size"`
+    // LVCreateOptions are extra arguments to pass to lvcreate
+    LVCreateOptions []string `json:"lvcreate-options"`
++   // Type of the volume to be created
++   Type string `json:"type",omitempty`
```

Comment:
What are the possible values of type ? Should this be an enum?

Reply:
- new type LVType which coerces to string values `thick` or `thin`

Reply:
Could you please remove `omitempty` option?
ref. https://github.com/topolvm/topolvm/pull/448#discussion_r834073781

Reply:
sure, I'll get this design PR refreshed at the earliest.